src/platform/platform_linux.cpp

```diff
@@ -29,10 +29,45 @@
 #include "logger/journald_logger.h"
 #include <disabled_update_prompt.h>
 
+#include <QStandardPaths>
+
+#include <cerrno>
+#include <cstring>
+#include <stdexcept>
+
 namespace mp = multipass;
 namespace mpl = multipass::logging;
 namespace mu = multipass::utils;
 
+namespace
+{
+constexpr auto autostart_filename = "multipass.gui.conditional-autostart.desktop";
+constexpr auto autostart_desktop_contents = "[Desktop Entry]\n"
+                                            "Name=Multipass\n"
+                                            "Exec=multipass.gui --autostarting\n"
+                                            "Icon=${SNAP}/meta/gui/multipass-gui.svg\n"
+                                            "Type=Application\n"
+                                            "Terminal=false\n"
+                                            "Categories=Utility;\n";
+} // namespace
+
+QString mp::platform::preliminary_gui_autostart_setup()
+{
+    const auto config_dir = QDir{QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation)};
+    const auto autostart_dir = QDir{config_dir.absoluteFilePath("autostart")};
+    const auto filepath = autostart_dir.absoluteFilePath(autostart_filename);
+
+    autostart_dir.mkpath(".");
+    QFile f{filepath};
+    if (!f.exists()) // assuming correct contents otherwise
```

Comment:
Should we be assuming correct contents? I wonder if this can be a symlink to a file in `$SNAP` instead?

BTW, if you wanted this generic, the `Icon=` path probably shouldn't point at `${SNAP}`?

Reply:
1. I initially thought of doing it along those lines, but later opted for this approach to avoid file ownership/permissions issues
2. Good catch. I am not sure what to do here though, any ideas? Maybe just dropping the icon altogether?

Reply:
1. We determine ownership of files in the snap, so that shouldn't be a problem.
2. Nothing comes to mind that will happen automagically… Snapcraft does rewrite the .desktop files, but only those that are actually bound to an `app` in `snapcraft.yaml`… We'd have to register a dummy "Multipass (autostart)" app to use that :(

Reply:
1. Let me break this down into a few points:
    1. "assuming correct contents": I could simply overwrite, if you wanted. That would kind of rule out manual tweaks, in a potentially confusing way (accepted at first, overwritten on 1st run). Maybe that's not a big problem and still something you could be happy with?
    2. I don't see how a link helps there: we can either overwrite it or it can still have/point-to anything
    3. I believe writing our own strings to files is something we already do in other places (cloud init stuff, libvirt xmls, apparmor profiles now)?
    4. given all this, I don't see much advantage in the symlink approach, and it would be more work before I can move on to other platforms.
    5. if you still want to do it, what ownership would you give the target? Currently the file is owned by the user.
2. OK, I'll just drop the icon

Reply:
1. Sure:
    1. I thought of overwriting, but that seems excessive. That's exactly where a symlink would help - it would point to a read-only location in `${SNAP}` so there's no modification that could be made. The right place to do modifications to this would be `~/.config/autostart`. In any case, people can remove the symlink and they're on their own.
    1. a symlink to a read-only file in `${SNAP}` is clearer and simpler than writing the file out (which may also lead to corruption of the file)
    1. the examples you gave are all per-instance (not sure about apparmor, if we could keep those in static files and link, we should)
    1. it would be less code to maintain in the end, though
    1. the file goes in the snap, and as long as it's world-readable, it should be fine
1. ACK

@gerboland @townsend2010 care to weigh in on this?

Reply:
1. 
    1. "people can remove the symlink and they're on their own" &ndash; generalizing, you can say "overwrite" in place of "remove"; the situation is parallel to a regular file
    2. I concede it may be clearer. Simpler, hmm maybe. I still don't see how it helps integrity: modification and corruption can happen in either case.
    3. ack
    4. ok, marginally: the string just moves to a separate file, but you replace open+write with a single link; error handling still needed though. Still more work now :stuck_out_tongue: 
    5. ofc, good point.

OK, I'll proceed with the other review items and wait for feedback from the others for this.

Reply:
Any corruption would have to happen on the packaging level, so a much bigger problem.

Also realized there's another advantage - handling packaging-specific differences (like `Icon=${SNAP}`) becomes much easier, if the packaging ships a static .desktop file to static location and we only link to it.

Reply:
> Any corruption would have to happen on the packaging level, so a much bigger problem.

packaging, deployment, reading... anyway, splitting hairs here

> another advantage - handling packaging-specific differences (like Icon=${SNAP}) becomes much easier

Sorry, I don't quite understand what you mean, could you explain?

Reply:
> Sorry, I don't quite understand what you mean, could you explain?

You were trying to find a way to keep the packaging differences away from the app itself, and having a packaged static file (that may have a different `Icon=` path, for example, or a different `Exec=` line) seems like a good approach to me. It makes any differentiation contained in packaging alone.

Reply:
Ah OK, I think I see what you mean.

Reply:
Just to summarize, as I stewed with this a bit again :)

I think the symlink approach makes for the best separation between app and packaging:
- app is only responsible for placing the link in `${XDG_CONFIG_HOME}/autostart`
- packaging is responsible for placing the actual `.desktop` file, applicable to the packaging, in the place where app is expecting it

The one scenario where this gets wonky is if the file needs to differ based on runtime properties of the system, rather than build time.

Reply:
OK, I'll do that then.

Reply:
@Saviq I starting doing this, but immediately stumbled into another hole: where to place the actual file in the first place (in the build dir)? Options I see:
- place it in `bin` -> I would still have to get the path to the current executable to find the desktop file (cwd can be something else entirely)
- somewhere relative to `bin` -> same problem + we'd have to keep the same relative path on `make install`
- use cmake to bake the file into the executable, to write it down later -> not much different from the current approach

So far autostart would work as long as `multipass.gui` was in the PATH. Now it would need to find another file. This seems to do away with the marginal simplicity gain of replacing "open+write" with "link". On the contrary, it feels like adding complexity for no tangible reason honestly.

As for packaging differences, there needn't be any. The `Icon` thing wouldn't work either way as you pointed above.

Reply:
Rather than baking the file contents into the executable in CMake, you can bake a "resources path" into it. Today it may only be there for the .desktop file, but we're bound to need more static resources with time.

Reply:
Having an absolute path in the executable? Is that what you're saying? Even if generated at build time, that would constrain the exe to a particular filesystem. And it would not address the `make install` question. If that is not what you're saying, I didn't get it.

Reply:
If the packaging system puts all the files in certain locations, I wouldn't expect that executable to be able to deal with changing those locations, unless that was documented behaviour.
Why would `make install` not be addressed by that? The `.desktop` file would be installed to a location based on install prefixes, and that location would also be where the executable would be looking for it.

An environment variable allowing to override that path on runtime is an option. I don't know of a standard environment variable that we could set to affect e.g. [`AppDataLocation`](https://doc.qt.io/qt-5/qstandardpaths.html#StandardLocation-enum), but I haven't read the code and even `XDG_` vars aren't mentioned in that documentation, so…

Reply:
Actually, it [seems to respect `XDG_DATA_DIRS`](https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/io/qstandardpaths_unix.cpp#n276).

Reply:
Sorry, I am getting lost in all this... I don't really see what you are proposing

Reply:
What about we offload that to a separate issue?

Reply:
You didn't like "absolute path in the executable", so another option is to use a standard path - `QStandardPaths` use the `XDG_DATA_DIRS` (note that's different to `_HOME`) environment variable to determine where data files for an application live. So instead of baking in an absolute path, we could use that, and rely on `QStandardPaths::AppDataLocation` (or one of the other ones, if applicable) to locate the static `.desktop` file.
Packaging would then be the one to make sure that either the file is indeed placed in the standard location (`/usr/share/…`, really), or `XDG_DATA_DIRS` is set as appropriate (in the snap case).

It doesn't help In the `make install` case unless you export `XDG_DATA_DIRS` to the right place in your install prefix.

I still think the link/unlink step is a much cleaner story for toggling autostart in the long run rather than writing out the file whole.

We're bound to need more resource files in the long run, so we can decide today how to make this happen.

Reply:
I don't want to merge code in that we'll have to deal with in another PR, because then we'll need to do a migration of some kind. We have the chance to set a precedent here, so let's do that.

Reply:
OK, so something like this? (please confirm or adapt):

- src:
    - new `resources` dir (`755` permissions)
    - new `.desktop` file there (`644` permissions)
- at `make`:
    - copy the file to a new `resources/multipass` dir
- at `make install`
    - find the first existing directory from `${XDG_DATA_DIRS}`, if not empty, `/usr/local/share/:/usr/share/` otherwise;
    - create a `multipass` subdir there
    - copy the file there
- in the snap:
    - use dump plugin to copy the `resources/multipass` dir
    - prepend the `resources` dir (not `multipass`) it to `XDG_DATA_DIRS`
- in the executable:
    - use [`QStandardPaths::locate`](https://doc.qt.io/qt-5/qstandardpaths.html#locate) to find the file (warning on failure)
    - create a link to it
- to test without installing
    - beside extending `PATH` we now need an additional step, namely adding the `resources` dir to `XDG_DATA_DIRS`

Except this won't really work because it assumes the app name `multipass` which is not the case in `multipass.gui`. So much for simplicity :joy:

Reply:
> * at `make install`
>   * find the first existing directory from `${XDG_DATA_DIRS}`, if not empty, `/usr/local/share/:/usr/share/` otherwise;
>   * create a `multipass` subdir there
>   * copy the file there

No, packaging only ever installs in one destination. Using [GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html) and `${CMAKE_INSTALL_DATADIR}` as destination in a call to [`install`](https://cmake.org/cmake/help/latest/command/install.html), or potentially [`configure_file`](https://cmake.org/cmake/help/latest/command/configure_file.html) when packaging differences needed.

> * in the snap:
>   * use dump plugin to copy the `resources/multipass` dir
>   * prepend the `resources` dir (not `multipass`) it to `XDG_DATA_DIRS`

Rather than the above, we just need to set `XDG_DATA_DIRS` to `${SNAP}/share`.

> Except this won't really work because it assumes the app name multipass which is not the case in multipass.gui. So much for simplicity :joy:

Not sure why `multipass` vs. `multipass.gui` matters here.
This moves responsibility to build/packaging time, over which we have much more control. The application only then needs to `locate` and link / unlink the `.desktop` file every time it wishes to enable/disable autostart. A much simpler operation than writing the file out.

Reply:
My quick stab at this: http://paste.ubuntu.com/p/9sDyYFypkS/

Reply:
I can see now what you meant with `multipass` vs. `multipass-gui`, that does put a proverbial stick between the spokes (not sure if that translates…). I'm looking into what we could do here.

Reply:
OK this is my proposal: http://paste.ubuntu.com/p/RJCPKQZwnh/

Reply:
> No, packaging only ever installs in one destination.

That is what I meant with "first existing directory". 

I like the standard `GNUInstallDirs` approach, but we have to hope it will always be included in `XDG_DATA_DIRS`. I suppose we can go with it and look back if anyone ever reports a failure.

> Rather than the above, we just need to set XDG_DATA_DIRS to ${SNAP}/share.

OK, so set rather than prepend and use the standard dir directly.

> This moves responsibility to build/packaging time, over which we have much more control. The application only then needs to locate and link / unlink the .desktop file every time it wishes to enable/disable autostart. A much simpler operation than writing the file out.

There is an elegance element in extracting the data from the app logic to packing. I am sensitive to that. I just don't think it is right to say "simpler" when it involves many more steps and possible failure points. I balance the pros and cons differently, so I see a different best approach. But of course I respect your preference. 

But my main point here was that the current approach is not quite enough yet. `QStandardPaths::AppDataLocation` will return something different in the two clients. Something else is still needed to make it work.

Reply:
OK, crossed comments and edits... 

It makes me a bit itchy to give two different programs the same "AppName". This would affect all QStandardPaths using the app name. How about we use `QStandardPaths::GenericDataLocation` instead:

*To obtain a path to store data to be shared with other applications, use QStandardPaths::GenericDataLocation*

We can append the `multipass` suffix manually when we want to share data. That allows both shared and app-specific data. For instance, imagine we want a separate settings file for the gui, we could still do that (not that it is needed, but just for the sake of example). Or if we wanted to place separate logs there or something...

Reply:
To me it really depends on the definition of AppName, but I suppose we're lacking that already.

I wanted to avoid coming up with our own version of `QStandardPaths::locate`, but maybe we really just need a `mp::StandardPaths` that has the properties we need. WFM.

Reply:
@Saviq, I just tested and `QStandardPaths::locate` works with dirs and relative paths:
```c++
qDebug() << QStandardPaths::locate(QStandardPaths::GenericDataLocation, "testdir", QStandardPaths::LocateDirectory);
qDebug() << QStandardPaths::locate(QStandardPaths::GenericDataLocation, "testdir/testfile", QStandardPaths::LocateFile);
```

Works :slightly_smiling_face:

Reply:
Oh ok, lacking in documentation again, then… Yeah, that works, then.

Reply:
@Saviq FYI I was doing some research into a possible alternative: install a symlink `multipass/bla.desktop` -> `multipass.gui/bla.desktop`. This would have allowed us to use a transparent app data path. 

It turned out to be a bit of a rabbit hole. There does not seem to be a nice way to achieve that. The best I got was something like

```cmake
set(autostart_filename multipass.gui.autostart.desktop)
set(install_dir ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_DATADIR})
set(autostart_main_dir ${install_dir}/multipass.gui)
set(autostart_secondary_dir ${install_dir}/multipass)
install(CODE "execute_process( \
        COMMAND ${CMAKE_COMMAND} -E create_symlink \
        \"\${DESTDIR}${autostart_main_dir}/${autostart_filename}\" \
        \"\${DESTDIR}${autostart_secondary_dir}/${autostart_filename}\"   \
    )"
        )
```

Notice that I escaped `\${DESTDIR}`, to interpret it only when actually doing `make install DESTDIR=foo`. Unfortunately this does not seem to work. The executed process apparently does not inherit the environment `make install` is executed with and `DESTDIR` always expands to an empty string.

Reverting back to the previous plan.

Reply:
@ricab yeah, I think the `::GenericDataLocation` approach is sane. The only caveat with it is that it's wonky when we want a writable location (for `multipass.conf` for example). So some wrapper around `QStandardPaths` may be in order at some point, so that we can easily and reliably determine a location common for all of Multipass, not each binary on its own.

Reply:
@Saviq ack.

Separate issue: I don't think the proposed change to snapcraft.yaml will work, for two reasons:

- `make install` actually installs it to `/usr/local` by default (see [this](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html))
- it would overwrite the entry `/var/lib/snapd/desktop`, which snapd presumably needs:

```console
$ snap run --shell multipass
$ echo $XDG_DATA_DIRS
/usr/local/share:/usr/share:/var/lib/snapd/desktop
```

Unless you see an issue with that, I am changing it to:

```yaml
XDG_DATA_DIRS: $SNAP/usr/local/share:$XDG_DATA_DIRS
```

Reply:
@ricab I confirmed this worked, though?
```
ls -lh /snap/multipass/current/usr/share/multipass
razem 512
-rw-r--r-- 1 root root 117 wrz  9 10:06 multipass.gui.autostart.desktop
```

Reply:
@ricab and we can override `snapd`'s value, it's not for its consumption, but for the snap's - and we know we won't need other app's desktop files. And if we need it, we can {pre,ap}pend.

Reply:
@ricab it's `CMAKE_INSTALL_PREFIX` that defaults to `/usr/local`, but it [gets overriden by Snapcraft](https://github.com/snapcore/snapcraft/blob/master/snapcraft/plugins/cmake.py#L116).

Reply:
@Saviq I see, I'll revert it then.